design/include-cleaner.md

```diff
@@ -0,0 +1,167 @@
+their definitions come from.
+
+After the process is complete, the declaration and definition locations will be
+collected and passed to the next stage.
+
+#### Marking the headers as used
+
+`SourceLocation` instances collected at the previous step will be converted to
+`FileID`s and deduplicated. In this stage, it is important to attribute the
+locations in some headers to their includes. Some of the `FileID`s correspond
+to non self-contained headers, meaning the user should actually include their
+parent rather than a header itself. Other important headers with this property
+are the ones manually marked as private through `IWYU pragma: private`. For
+them, the user explicitly asks includers to consider the public header.
+
+After the responsible headers are collected, they only step left is producing
+diagnostics for unused headers.
+
+#### Issuing warnings
+
+After most of the work is done, Include Cleaner needs to decide which headers
+are not used in the main file. All inclusions are scanned and checked for
+containing `IWYU pragma: keep`; if they do not, they are not used and will be
+warned about.
+
+### IWYU pragmas
+
+IWYU tool offers a set of
+[pragmas](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUPragmas.md),
+the most notable of which are:
+
+- `IWYU pragma: keep` indicates the inclusion should not be removed from the
+  main file
+- `IWYU pragma: private` indicates that the current header is an implementation
+  detail and another one should be included instead
+
+## Future plans and Limitations
+
+### Umbrella headers
+
+The "umbrella" headers are re-exporting the implementation headers for public
+use, which is a common practice in Open Source projects. A notable example of
+this is Googletest: the `gtest/gtest.h` top-level header does not contain any
+definitions: it includes a number of "internal" headers that are not
+recommended to users. The users should always write `#include
+"gtest/ghtest.h"`. The way to propagate that information to Include Cleaner is
+using `// IWYU pragma: private, include "public.h"` in your `private.h` header
+that is being exported.
+
+### Standard library
+
+By default, Include Cleaner will not diagnose headers from the Standard
+Library. Standard Library headers support is not complete yet (due to the
+macros and the fact that a symbol is allowed to come from multiple headers) but
+you can enable this unstable feature through passing `--include-cleaner-stdlib`
+flag to clangd invocation.
+
+### Inserting Includes
+
+The complete version of Include Cleaner will not only warn about unused
+headers, but also provide a way to include used headers directly, not through a
+chain of transitive includes.
+
+### Template type aliases
+
+Include Cleaner does not currently support the templated type aliases because
+Clang AST does not record the fact that expanded type was reached through a
+type alias in cases like this:
+
+```c++
+// vec.h
```

Comment:
i think there's actually a dedicated node for representing the `Vec` in this case (a `TypeAliasTemplateDecl`).
Can we give the more problematic using decl example:
```
namespace ns {
using std::vector;
}
```
now depending on `ns::vector` will only require header defining `std::vector`.